test/unit/core/type_traits/template_inspection_test.cpp

```diff
@@ -34,6 +40,29 @@ TEST(template_inspect, transfer_template_args_onto_t)
     EXPECT_TRUE((std::is_same_v<t, std::tuple<int, char, double>>));
 }
 
+TEST(template_inspect, transfer_template_args_onto_with_constraint)
+{
+    // This test ensures that transfer_template_args_onto uses internally only
+    // type declarations and not type instantiations
+
+    using bar_char = constraint_bar<char>; // this is fine, even though it contains a static_assert
```

Comment:
how? Is char cast to int or does the static_assert only hit if you instantiate bar_char?

Reply:
yeah that's the point :)

* `constraint_bar<char>{}` would static_assert.
* `std::declval<constraint_bar<char>>()` would static_assert.
* `using bar_char = constraint_bar<char>;` does not static_assert
* `typename constraint_bar<char>;` in concepts does not static_assert

 the expression `using bar_char = constraint_bar<char>;` just (forward) declares and does not evaluate the class body, so the `static_assert` will not be evaluated.

Reply:
Ah, here it is explained as well. 👍